ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -994,6 +999,40 @@ module Winbind = struct
       Printf.sprintf "%s-%s" prefix suffix
     else
       hostname
+
+  let rotate_machine_password_task_name = "Rotating machine password"
+
+  let rotate_machine_password () =
+    let now = Unix.time () in
+    let now_str = string_of_float now in
+    try
+      let machine_pwd_last_change_time =
+        (get_domain_info_from_db ()).machine_pwd_last_change_time
+      in
+      match machine_pwd_last_change_time with
+      | Some time when now < time +. !Xapi_globs.winbind_machine_pwd_timeout ->
+          ()
+      | _ ->
+          debug "%s started at %s" rotate_machine_password_task_name now_str ;
+          Helpers.call_script net_cmd
+            ["changetrustpw"; "--kerberos"]
```

Comment:
I don't think this is guaranteed to use the same KDC as XAPI, and if it doesn't then there will be a time where the machine password will be out of date: the KDC which 'net changetrustpw' used will have the new machine password and the other KDCs will have the old one. So depending which KDC you will use, you will either succeed further operations or fail because the machine password is wrong and out of date.
And if the KDCs are geographically distributed and slow to synchronize we know it can take 15m-30m until a password change update is propagates to all KDCs.

Is it possible to tell winbind here which KDC to prefer, so that at least we have a higher chance of this to succeed? Otherwise you'll get a 15m-30m window where authentications randomly fail or succeed.

Reply:
I think both DC and samba/keytab keep the latest 2 version of password, so it is okay as long as we does not changeturst often (default to 2 weeks), then DC will sync the password .

Reply:
OK, maybe it works differently for machine passwords, changing my user password used to be problematic.